boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,355 @@
+
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_WriteRegister(SI_GCR_ADDR, register_value, e_fuse));
+
+    return EXIT_CODE_OK;
+}
+
+ExitCode Io_Efuse_ExitFailSafeMode(struct Efuse_Context *const e_fuse)
+{
+    // Set WDIN bit for next write
+    // 1_1_00000_00000_0000
+    e_fuse->wdin_bit_to_set = true;
+
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_WriteRegister(SI_STATR_0_ADDR, 0x0000, e_fuse));
+
+    // Disable the watchdog timer
+    RETURN_CODE_IF_EXIT_NOT_OK(
+        Io_Efuse_WriteRegister(SI_GCR_ADDR, GCR_CONFIG, e_fuse));
+
+    // Check if the the efuse is still in fail-safe mode
+    if (Io_Efuse_IsEfuseInFailSafeMode(e_fuse))
+    {
+        return EXIT_CODE_UNIMPLEMENTED;
+    }
+
+    return EXIT_CODE_OK;
+}
+
+bool Io_Efuse_IsEfuseInFaultMode(const struct Efuse_Context *const e_fuse)
+{
+    return HAL_GPIO_ReadPin(e_fuse->fsb_port, e_fuse->fsb_pin);
+}
+
+bool Io_Efuse_IsEfuseInFailSafeMode(const struct Efuse_Context *const e_fuse)
+{
+    return HAL_GPIO_ReadPin(e_fuse->fsob_port, e_fuse->fsob_pin);
+}
+
+void Io_Efuse_DelatchFaults(const struct Efuse_Context *const e_fuse)
+{
+    // Delatch the latchable faults by alternating channel 0's and channel 1's
+    // input pins high->low->high.
+    HAL_GPIO_WritePin(
+        e_fuse->channel0_port, e_fuse->channel0_pin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(
+        e_fuse->channel0_port, e_fuse->channel0_pin, GPIO_PIN_RESET);
+    HAL_GPIO_WritePin(
+        e_fuse->channel0_port, e_fuse->channel0_pin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(
+        e_fuse->channel1_port, e_fuse->channel1_pin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(
+        e_fuse->channel1_port, e_fuse->channel1_pin, GPIO_PIN_RESET);
+    HAL_GPIO_WritePin(
+        e_fuse->channel1_port, e_fuse->channel1_pin, GPIO_PIN_SET);
+}
+
+ExitCode Io_Efuse_WriteRegister(
+    uint8_t                     register_address,
+    uint16_t                    register_value,
+    struct Efuse_Context *const e_fuse)
+{
+    uint16_t serial_input_data = 0x0000U;
+
+    // Place the register address into bits 10->13
+    serial_input_data = (uint16_t)(
+        serial_input_data |
+        ((register_address & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT));
+
+    // Place register value to be written into bits 0->8
+    serial_input_data =
+        (uint16_t)(serial_input_data | (register_value & EFUSE_SI_DATA_MASK));
+
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        SET_BIT(serial_input_data, WATCHDOG_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(serial_input_data, WATCHDOG_BIT);
+    }
+
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+    Io_Efuse_CalculateParityBit(&serial_input_data);
+
+    return Io_Efuse_WriteToEfuse(
+        &serial_input_data, e_fuse->hspi, e_fuse->nss_port, e_fuse->nss_pin);
```

Comment:
Honestly I remember recommending serial_input_data but I think tx_data is a better name

Reply:
top 10 anime betrayals

Reply:
👺

Reply:
Changed

Reply:
I still don't see changes

Reply:
didn't push lol

Reply:
local changes moment